File: src/main/java/org/operon/tutorials/Tutorial.java
package org.operon.tutorials;

import java.util.List;

import javax.persistence.Id;

//@Document //(collection="tutorials")
public class Tutorial
{
	@Id private String id;
	private String name=null;
	private Integer number;
	private String title="";

	private ItemCollection<Page> pages=new ItemCollection<Page>();
	private ItemCollection<Question> questions=new ItemCollection<Question>();
	private ItemCollection<Image> images=new ItemCollection<Image>();
	private ItemCollection<Link> links=new ItemCollection<Link>();
	private ItemCollection<Flash> flashs=new ItemCollection<Flash>();
	
	public Tutorial(){}
	
	public Tutorial(String name)
	{
		this.name=name;
	}

	public String getId(){return this.id;}
	public void setId(final String id){this.id=id;}
	
	public String getName(){return this.name;}
	public void setName(String name){this.name=name;}

	public Integer getNumber(){return this.number;}
	public void setNumber(Integer number){this.number=number;}

	public String getTitle(){return this.title;}
	public void setTitle(String title){this.title=title;}
	public List<Page> getPages(){return this.pages;}
	public List<Image> getImages(){return this.images;}	
	public List<Link> getLinks(){return this.links;}
	public List<Flash> getFlashs(){return this.flashs;}
	public List<Question> getQuestions(){return this.questions;}
	
	public Page add(Page page)
	{
		return this.pages.addItem(page);
	}
	
	public Question add(Question question)
	{
		return this.questions.addItem(question);
	}
	
	public Image add(Image image)
	{
		return this.images.addItem(image);
	}
	
	public Link add(Link link)
	{
		return this.links.addItem(link);
	}
	
	public Flash add(Flash flash)
	{
		return this.flashs.addItem(flash);
	}
	
	public Page findOrCreatePage(final String page_id)
	{
		Page page=this.pages.findItem(page_id);
		if (page==null)
			return add(new Page(page_id));
		return page;
	}
	
	public Question findOrCreateQuestion(String question_id)
	{
		Question question=this.questions.findItem(question_id);
		if (question==null)
			return add(new Question(question_id));
		return question;
	}
	
	public Image findOrCreateImage(String image_id)
	{
		Image image=this.images.findItem(image_id);
		if (image==null)
			return add(new Image(image_id));
		return image;
	}
	
	public Link findOrCreateLink(String link_id)
	{
		Link link=this.links.findItem(link_id);
		if (link==null)
			return add(new Link(link_id));
		return link;
	}
	
	public Flash findOrCreateFlash(String flash_id)
	{
		Flash flash=this.flashs.findItem(flash_id);
		if (flash==null)
			return add(new Flash(flash_id));
		return flash;
	}
}

File: dev/operon2/js/models/tutorial.js
/*global Ext */
Ext.regModel('Tutorial',
{
	idProperty: '_id',
	fields:
	[
	 	{name: '_id'},
	 	{name: 'name'},
	 	{name: 'number', type: 'int'},
	 	{name: 'title'}
	],
	hasMany:
	[
	 	{model: 'Page', name : 'pages'},
	 	{model: 'Question', name : 'questions'},
		{model: 'Image', name : 'images'},
		{model: 'Link', name : 'links'},
		{model: 'Flash', name : 'flashs'}
	],
	proxy:
	{
		 type: 'rest',
		 url: '/json',
		 reader: 'json'
	},
 	questionnumber: 1,
	imagenumber: 1,
	linknumber: 1,
	flashnumber: 1,
	
	start: function()
	{
		this.toc=new operon.tutorials.Toc(this);
		this.score=new operon.tutorials.Score(this);
		this.indexResources();
		
		// render
		var self=this;
		var backbutton = Ext.get('backbutton');
		backbutton.on('click',function(evt,t,o)
		{
			self.previousPage();
		});
		
		var nextbutton = Ext.get('nextbutton');
		nextbutton.on('click',function(evt,t,o)
		{
			self.nextPage();			
		});
		
		document.title=this.get('number')+'. '+this.get('title');
		Ext.core.DomHelper.overwrite('tutorial_title',document.title);

		//var page=this.pages().first();
		var page=this.getPageByNumber(6); //hack
		this.gotoPage(page);
	},

	gotoPage:function(page)
	{
		if (this.curpage)
			this.curpage.current=false;
		this.curpage=page;
		page.set('current',true);
		page.set('enabled',true);
		page.update();
		this.toc.update();
	},
	
	indexResources:function()
	{
		var self=this;
		var number=1;
		this.pages().each(function(page)
		{
			page.set('number',number);
			page.indexResources(self);
			page.prevpage=number-1;
			page.nextpage=number+1;
			number++;
		});
		
		var page=this.pages().first();
		page.showback=false;
		page.prevpage=null;
		
		page=this.pages().last();
		page.shownext=false;
		page.nextpage=null;

		// sort resources
//		this.questions.sort(this.sortByNumber);
//		this.images.sort(this.sortByNumber);
//		this.links.sort(this.sortByNumber);
//		this.flashs.sort(this.sortByNumber);
	},

	getPageById:function(page_id)
	{
		return vardb.util.Utils.getArrayItem(this.pages(),"id",page_id);
	},
	
	getPageByNumber:function(number)
	{
		return vardb.util.Utils.getArrayItem(this.pages(),"number",number);
	},
	
	getQuestion:function(question_id)
	{
		return vardb.util.Utils.getArrayItem(this.questions(),"id",question_id);
	},
	
	getImage:function(image_id)
	{
		return vardb.util.Utils.getArrayItem(this.images(),"id",image_id);
	},
	
	getLink:function(link_id)
	{
		return vardb.util.Utils.getArrayItem(this.links(),"id",link_id);
	},
	
	getFlash:function(flash_id)
	{
		return vardb.util.Utils.getArrayItem(this.flashs(),"id",flash_id);
	},

	gotoPageId:function(page_id)
	{
		var page=this.getPageById(page_id);
		this.gotoPage(page);
	},
	
	gotoPageNumber:function(page_id)
	{
		var page=this.getPageByNumber(page_id);
		this.gotoPage(page);
	},
	
	previousPage:function()
	{
		var page=this.curpage;
		if (!page.prevpage)
			{return;}
		this.gotoPageNumber(page.prevpage);
	},
	
	nextPage:function()
	{
		var page=this.curpage;
		if (!page.nextpage)
			{return;}
		this.gotoPageNumber(page.nextpage);
	},

	submitAnswer:function(question_id,choice_id)
	{
		//alert('question_id='+question_id+', choice_id='+choice_id);
		var question=this.getQuestion(question_id);
		if (!choice_id)
			{return;}
		question.submitAnswer(choice_id);
	},
		
	getScore:function()
	{
		return this.score;
	},
	
	sortByNumber:function(a,b)
	{
		return a.number-b.number;
	},
	
	getResource:function(type, name)
	{
		if (type==='question')
			{return this.getQuestion(name);}
		else if (type==='image')
			{return this.getImage(name);}
		else if (type==='link')
			{return this.getLink(name);}
		else if (type==='flash')
			{return this.getFlash(name);}
	},
	
	printTutorial:function()
	{
		var win=this.openWindow('about:blank');
		
		var index,page;
		var str='<h1>'+this.number+'. '+this.tutorial_title+'</h1>\n';
		for (index=0;index<this.pages.length;index++)
		{
			page=this.pages[index];
			if (page.complete)
				{str+=page.printPage();}
		}
		win.document.write(str);
		win.document.close();
	},
	
	openWindow:function(url,title,options)
	{
		title=title || url;
		options=options || 'scrollbars=1,resizable=1';
		var win=window.open(url,title,options);
		if (!win)
		{
			win = window.open('',title, options);
			win.location.href = url;
		}	
		if (window.focus)
			{win.focus();}
		return win;
	}
});
